tests/kernel/wave/wave_sim_test.py

```diff
@@ -210,3 +210,192 @@ def repeat(acc: tkl.Register[M, N, tkl.f32]) -> tkl.Register[M, N, tkl.f32]:
+
+    @wave_sim(constraints)
+    def transpose(
+        a: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f32],
+        c: tkl.Memory[N, M, ADDRESS_SPACE, tkl.f32],
+    ):
+        a_reg = tkw.read(a, mapping=mapping, elements_per_thread=LOAD_ELEMS_PER_THREAD)
+        tkw.write(a_reg, c, elements_per_thread=STORE_ELEMS_PER_THREAD)
+
+    a = torch.randn(128, 256, dtype=torch.float32)
+    c = torch.zeros(256, 128, dtype=torch.float32)
+    transpose(a, c)
+    assert_allclose(c, a.T)
+
+
+def test_transpose_2():
+    # Input sizes
+    M = tkl.sym.M
+    N = tkl.sym.N
+    # Workgroup tile sizes
+    BLOCK_M = tkl.sym.BLOCK_M
+    BLOCK_N = tkl.sym.BLOCK_N
+    # Address space (for GPU, shared(1) or global(0))
+    ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+    # Other hyperparameters
+    LOAD_ELEMS_PER_THREAD = tkl.sym.LOAD_ELEMS_PER_THREAD
+    STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+    # Expose user-constraints
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+
+    constraints += [
+        tkw.HardwareConstraint(threads_per_wave=64, waves_per_block=(1, 1, 1))
+    ]
+
+    i = tkw.IndexMapping.iterator(0)
+    j = tkw.IndexMapping.iterator(1)
+    mapping = tkw.IndexMapping(num_dims=2, inputs={N: i, M: j}, outputs={N: i, M: j})
+
+    @wave_sim(constraints)
+    def transpose(
+        a: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f32],
+        c: tkl.Memory[N, M, ADDRESS_SPACE, tkl.f32],
+    ):
+        a_reg = tkw.read(a, elements_per_thread=LOAD_ELEMS_PER_THREAD)
+        tkw.write(
+            a_reg,
+            c,
+            mapping=mapping,
+            elements_per_thread=STORE_ELEMS_PER_THREAD,
+        )
+
+    a = torch.randn(128, 256, dtype=torch.float32)
+    c = torch.zeros(256, 128, dtype=torch.float32)
+    transpose(a, c)
+    assert_allclose(c, a.T)
+
+
+@pytest.mark.parametrize("n", [1, 2, 4])
+@pytest.mark.parametrize("c", [1, 3, 10])
+@pytest.mark.parametrize("nf", [1, 2, 8])
+@pytest.mark.parametrize("stride", [1, 2, 3])
+def test_igemm_conv(n, c, nf, stride):
+    h, w = 5, 5  # Image.
+    cf, hf, wf = c, 2, 2  # Filters.
+    x = torch.randn(n, c, h, w, dtype=torch.float32)
+    we = torch.randn(nf, cf, hf, wf, dtype=torch.float32)
+
+    padding = 0  # TODO: only pad=0 is supported for now
+    convRef = torch.nn.Conv2d(c, nf, hf, stride=stride, padding=padding, bias=False)
+    convRef.weight = torch.nn.Parameter(we)
+    out_ref = convRef(x).detach()
+
+    sym = tkl.sym
+    N, C, H, W = sym.N, sym.C, sym.H, sym.W
+    NF, HF, WF = sym.NF, sym.HF, sym.WF
+
+    H_OUT = (H + 2 * padding - HF) // stride + 1
+    W_OUT = (W + 2 * padding - WF) // stride + 1
+    SZ_OUT = H_OUT * W_OUT
+
+    K = HF * WF * C
+    M = SZ_OUT * N
+
+    i = tkw.IndexMapping.iterator(0)
+    j = tkw.IndexMapping.iterator(1)
+
+    x_mapping = tkw.IndexMapping(

```

Comment:
Just confirming here that the reason we need to specify num_dims is because in general we cannot deduce which variables are iterators and which ones are not?

Reply:
Technically, we can remove it and just enumerate all iterators both in `input` and `output` instead (as each iterator must be used at least once), but I think it's more user readable this way. (Also, after some thought, `num_iterators` will be a better name).